spark/src/main/scala/org/apache/spark/sql/delta/SnapshotManagement.scala

```diff
@@ -288,17 +307,92 @@ trait SnapshotManagement { self: DeltaLog =>
+   * @param checkpointVersionToUse - underlying checkpoint version to use in Snapshot, -1 if no
+   *                               checkpoint is used.
+   * @return Returns a list of deltas/compacted-deltas which can be used to construct the
+   *         [[LogSegment]] instead of `deltasAfterCheckpoint`.
+   */
+  protected def useCompactedDeltasForLogSegment(
+      deltasAndCompactedDeltas: Seq[FileStatus],
+      deltasAfterCheckpoint: Array[FileStatus],
+      latestCommitVersion: Long,
+      checkpointVersionToUse: Long): Array[FileStatus] = {
+
+    val selectedDeltas = mutable.ArrayBuffer.empty[FileStatus]
+    var highestVersionSeen = checkpointVersionToUse
+    val commitRangeCovered = mutable.ArrayBuffer.empty[Long]
+    // track if there is at least 1 compacted delta in `deltasAndCompactedDeltas`
+    var hasCompactedDeltas = false
+    for (file <- deltasAndCompactedDeltas) {
+      val (startVersion, endVersion) = file match {
+        case CompactedDeltaFile(_, startVersion, endVersion) =>
+          hasCompactedDeltas = true
+          (startVersion, endVersion)
+        case DeltaFile(_, version) =>
+          (version, version)
+      }
+
+      if (highestVersionSeen < startVersion && endVersion <= latestCommitVersion) {
+        commitRangeCovered.appendAll(startVersion to endVersion)
+        selectedDeltas += file
+        highestVersionSeen = endVersion
+      }
+    }
+    // If there are no compacted deltas in the `deltasAndCompactedDeltas` list, return from this
+    // method.
+    if (!hasCompactedDeltas) return deltasAfterCheckpoint
+    // Validation-1: Commits represented by `compactedDeltasToUse` should be unique and there must
+    // not be any duplicates.
+    val coveredCommits = commitRangeCovered.toSet
+    val hasDuplicates = (commitRangeCovered.size != coveredCommits.size)
+
+    // Validation-2: All commits from (CheckpointVersion + 1) to latestCommitVersion should be
+    // either represented by compacted delta or by the delta.
+    val requiredCommits = (checkpointVersionToUse + 1) to latestCommitVersion
+    val missingCommits = requiredCommits.toSet -- coveredCommits
+    if (!hasDuplicates && missingCommits.isEmpty) return selectedDeltas.toArray
+
+    val eventData = Map(
+      "deltasAndCompactedDeltas" -> deltasAndCompactedDeltas.map(_.getPath.getName),
+      "deltasAfterCheckpoint" -> deltasAfterCheckpoint.map(_.getPath.getName),
+      "latestCommitVersion" -> latestCommitVersion,
+      "checkpointVersionToUse" -> checkpointVersionToUse,
+      "hasDuplicates" -> hasDuplicates,
+      "missingCommits" -> missingCommits
+    )
+    recordDeltaEvent(
+      deltaLog = this,
+      opType = "delta.getLogSegmentForVersion.compactedDeltaValidationFailed",
+      data = eventData)
+    if (Utils.isTesting) {
+      assert(false, s"Validation around Compacted deltas failed while creating Snapshot. " +
+        s"[${JsonUtils.toJson(eventData)}]")
+    }
```

Comment:
What does this code do? It seems to fail unconditionally in testing, but it's not obvious why validation has failed at this point?

Reply:
adding comment before the `val eventData`.